src/Agent.Sdk/SecretMasking/SecretMasker.cs

```diff
@@ -406,7 +97,319 @@ public void AddRegex(string pattern, RegexOptions options)
+            }
+        }
+    }
+
+
+    public ISecretMasker Clone() => new SecretMasker(this);
+
+    public void Dispose()
+    {
+        Dispose(true);
+        GC.SuppressFinalize(this);
+    }
+
+    protected virtual void Dispose(bool disposing)
+    {
+        if (disposing)
+        {
+            m_lock?.Dispose();
+            m_lock = null;
+        }
+    }
+
+    public String MaskSecrets(String input)
+    {
+        if (String.IsNullOrEmpty(input))
+        {
+            return String.Empty;
+        }
+
+        var secretPositions = new List<Replacement>();
+
+        // Read section.
+        try
+        {
+            m_lock.EnterReadLock();
+
+            var stopwatch = Stopwatch.StartNew();
+
+            // Get indexes and lengths of all substrings that will be replaced.
+            // We also persist the generated replacement tokens for secrets. In
+            // some cases, this data will be put in the telemetry stream.
+            foreach (RegexSecret regexSecret in m_regexSecrets)
+            {
+                foreach (Replacement replacement in regexSecret.GetReplacements(input))
+                {
+                    ReplacementTokens.Add(replacement.Token);
+                    secretPositions.Add(replacement);
+                }
+            }
+
+            foreach (ValueSecret valueSecret in m_valueSecrets)
+            {
+                secretPositions.AddRange(valueSecret.GetPositions(input));
+            }
+
+            ElapsedMaskingTime += stopwatch.ElapsedTicks;
+        }
+        finally
+        {
+            if (m_lock.IsReadLockHeld)
+            {
+                m_lock.ExitReadLock();
+            }
+        }
+
+        // Short-circuit if nothing to replace.
+        if (secretPositions.Count == 0)
+        {
+            return input;
+        }
+
+        // Merge positions into ranges of characters to replace.
+        List<Replacement> replacementPositions = new List<Replacement>();
+        Replacement currentReplacement = null;
+        foreach (Replacement secretPosition in secretPositions.OrderBy(x => x.Start))
+        {
+            if (currentReplacement == null)
+            {
+                currentReplacement = new Replacement(copy: secretPosition);
+                replacementPositions.Add(currentReplacement);
+            }
+            else
+            {
+                if (secretPosition.Start <= currentReplacement.End)
+                {
+                    // Overlap
+                    currentReplacement.Length = Math.Max(currentReplacement.End, secretPosition.End) - currentReplacement.Start;
+                }
+                else
+                {
+                    // No overlap
+                    currentReplacement = new Replacement(copy: secretPosition);
+                    replacementPositions.Add(currentReplacement);
+                }
+            }
+        }
+
+        // Replace
+        var stringBuilder = new StringBuilder();
```

Comment:
>StringBuilder [](http://example.com/codeflow?start=32&length=13)

StringBuilder instance construction is pretty expensive, we should consider a thread local field for a dedicated string builder instance (and set its length to 0 in this code path). A future change we can look at, maybe, to test our flighting/performance telemetry-based testing).